lib/db/observed.go

```diff
@@ -0,0 +1,265 @@
+	key, err := db.keyer.GeneratePendingFolderKey(nil, device[:], []byte(id))
+	if err == nil {
+		of := ObservedFolder{
+			Time:  time.Now().Round(time.Second),
+			Label: label,
+		}
+		bs, err := of.Marshal()
+		if err == nil {
+			err = db.Put(key, bs)
+		}
+	}
+	return err
+}
+
+// PendingFolders drops any invalid entries from the database as a side-effect.
+func (db *Lowlevel) PendingFolders() (map[string]map[protocol.DeviceID]ObservedFolder, error) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingFolder})
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Release()
+	res := make(map[string]map[protocol.DeviceID]ObservedFolder)
+	for iter.Next() {
+		keyDev, ok := db.keyer.DeviceFromPendingFolderKey(iter.Key())
+		if !ok {
+			if err := db.deleteInvalidPendingFolder(iter.Key()); err != nil {
+				return nil, err
+			}
+			continue
+		}
+		// Here we expect the length to match, coming from the device index.
+		deviceID := protocol.DeviceIDFromBytes(keyDev)
+		if err := db.collectPendingFolder(iter.Key(), deviceID, res); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+// PendingFoldersForDevice drops any invalid entries from the database as a side-effect.
+func (db *Lowlevel) PendingFoldersForDevice(device protocol.DeviceID) (map[string]map[protocol.DeviceID]ObservedFolder, error) {
+	prefixKey, err := db.keyer.GeneratePendingFolderKey(nil, device[:], nil)
+	if err != nil {
+		return nil, err
+	}
+	iter, err := db.NewPrefixIterator(prefixKey)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Release()
+	res := make(map[string]map[protocol.DeviceID]ObservedFolder)
+	for iter.Next() {
+		if err := db.collectPendingFolder(iter.Key(), device, res); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+func (db *Lowlevel) collectPendingFolder(key []byte, device protocol.DeviceID, res map[string]map[protocol.DeviceID]ObservedFolder) error {
+	bs, err := db.Get(key)
+	if err != nil {
+		return err
+	}
+	var of ObservedFolder
+	folderID := string(db.keyer.FolderFromPendingFolderKey(key))
+	if len(folderID) < 1 {
+		return db.deleteInvalidPendingFolder(key)
+	}
+	if err := of.Unmarshal(bs); err != nil {
+		return db.deleteInvalidPendingFolder(key)
+	}
+	if _, ok := res[folderID]; !ok {
+		res[folderID] = make(map[protocol.DeviceID]ObservedFolder)
+	}
+	res[folderID][device] = of
+	return nil
+}
+
+// deleteInvalidPendingFolder logs a warning message before dropping the given entry.
+// That's the only possible "repair" measure and appropriate for the importance of pending
+// entries.  They will come back soon if still relevant.
+func (db *Lowlevel) deleteInvalidPendingFolder(key []byte) error {
+	l.Warnf("Invalid pending folder entry, deleting from database: %x", key)
+	err := db.Delete(key)
+	return err
+}
+
+// Set of devices for which pending folders are allowed, but not specific folder IDs
+type DropListObserved map[protocol.DeviceID]map[string]struct{}
```

Comment:
> I don't understand the comment in relation to the type, nor really the type name (is it a noun or a verb?). This is a `device -> set of folders` thing so it surely deals with specific folder IDs?

I'm not very happy with the name either. It is "the drop list" we have been repeatedly discussing in this PR, and I added the `Observed` suffix just because it IS exported and that's the name of the containing file.

> Also given this is an exported type that someone outside is going to construct (I guess?) `map[protocol.DeviceID][]string` would make more sense I think.

The two-level nested map is designed this way for fast look-ups when we need to make a decision. Based on @imsodin's feedback, we try to iterate over the relevant DB entries only once and do lookups within this map in RAM instead of iterating over the map (or slice) and repeatedly looking up DB keys, which is presumably much slower.

> If the intention is not for the consumer to fiddle with the structure I think it would be better encapsulated as a struct

It is constructed through the `NewDropListObserved()` function solely and I took care to make the consumer agnostic of it being a map.  I don't understand what you mean with encapsulating it as a struct.  It still needs an exported name, doesn't it?

> Is this really a `KnownDeviceFolderSet` (a set of known devices and their corresponding folders)?

Not quite. I'd made a decision matrix to illustrate the semantics:
![pending-cleanup_2020-06-04](https://user-images.githubusercontent.com/6996887/83815565-bf33dc00-a6c0-11ea-9838-cde61a4a4140.png)

The hard part is making it drop anything that we cannot enumerate because it is only mentioned in the DB, like for pending folders of removed devices.  I've been through several designs for collecting that info, but it always ended up very verbose and distracting in the model methods, so it's now encapsulated here.  Using a common type for `CleanPendingFolders()` and `CleanPendingDevices()` avoids building two sets while going through the configuration object. It now only needs the addition of ignored device IDs before calling the latter function, otherwise reusing the contents of the first.

I see how that whole cleanup thing looks drastically more complicated than what used to be done within the configuration object.  It was unexpectedly *the one* complicated problem to solve compared to the rest of this PR, so I'd have liked a simpler solution myself.  But after much discussion, this is the most elegant way we came up with (avoiding DB performance overhead).  Maybe @imsodin can share his perspective on the justification?